pkg/controllers/securitypolicy_controller.go

```diff
@@ -17,37 +21,106 @@ import (
+	GCInterval    = 60 * time.Second
+	finalizerName = "securitypolicy.nsx.vmware.com/finalizer"
 )
 
 // SecurityPolicyReconciler reconciles a SecurityPolicy object
 type SecurityPolicyReconciler struct {
-	client.Client
+	Client  client.Client
 	Scheme  *runtime.Scheme
 	Service *services.SecurityPolicyService
 }
 
 func (r *SecurityPolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-
 	obj := &v1alpha1.SecurityPolicy{}
-	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
-		log.Error(err, "unable to fetch object")
-		return ctrl.Result{}, err
+	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch security policy CR")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	log.Info("reconciling securitypolicy CR", "securitypolicy", req.NamespacedName)
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		if err := r.Service.CreateOrUpdateSecurityPolicy(obj); err != nil {
+			log.Error(err, "failed to create or update security policy", "securitypolicy", req.NamespacedName)
+			return ctrl.Result{}, err
+		}
+		if !controllerutil.ContainsFinalizer(obj, finalizerName) {
+			controllerutil.AddFinalizer(obj, finalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		if containsString(obj.GetFinalizers(), finalizerName) {
+			if err := r.Service.DeleteSecurityPolicy(obj.UID); err != nil {
+				return ctrl.Result{}, err
+			}
 
-	if err := r.Service.CreateOrUpdateSecurityPolicy(obj); err != nil {
-		log.Error(err, "failed to create or update security policy", "securitypolicy", req.NamespacedName)
-		return ctrl.Result{}, err
+			controllerutil.RemoveFinalizer(obj, finalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
 	}
 
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *SecurityPolicyReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func containsString(source []string, target string) bool {
+	for _, item := range source {
+		if item == target {
+			return true
+		}
+	}
+	return false
+}
+
+func (r *SecurityPolicyReconciler) setupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&v1alpha1.SecurityPolicy{}).
 		Complete(r)
 }
+
+// Start setup manager and launch GC
+func (r *SecurityPolicyReconciler) Start(mgr ctrl.Manager) error {
+	err := r.setupWithManager(mgr)
+	if err != nil {
+		return err
+	}
+
+	go r.GarbageCollector()
+	return nil
+}
+
+func (r *SecurityPolicyReconciler) GarbageCollector() {
+	ctx := context.Background()
+	log.V(1).Info("garbage collector started")
+	for {
+		nsxPolicySet, err := r.Service.ListSecurityPolicy()
+		if err != nil {
+			log.Error(err, "failed to get security policy from local store")
+		}
```

Comment:
1. you added the error handling, but in the definition of `ListSecurityPolicy`, I saw the second return value is always `nil`, it seems that L103~105 cannot be invoked?
2. here we didn't get the nsx security policy from local store, should we continue the logic of L106~123?
3. "get security policy" -> "get NSX security policy"? for readability.

Reply:
remove L103~L105
 done